### DISCUSSION TITLE
Can not find image XObject /Matrix data

### DISCUSSION THREAD

**Discussion by jianfan2012** [py-pdf/pypdf#2163] [Show and tell]
## Explanation
I have a pdf file with couple of images, when I extracted the image, the images are upside down, I was able to use PyMUPDF to get the transformation matrix , but with pypdf the image XObject does not contain the transformation matrix for those images. because of license issue, we can not use PyMUPDF package, is there way to see the image transformation matrix using pypdf package?

## Code Example
from pypdf import PdfReader
pdf_file="./jpegcompress.pdf"
doc = PdfReader(pdf_file)
p0=doc.pages[0]
print(p0['/Resources']['/XObject']['/Im1'])
{'/BitsPerComponent': 8, '/ColorSpace': IndirectObject(97, 0, 140707430812304), '/Decode': [0.0, 255], '/Filter': '/FlateDecode', '/Height': 560, '/Name': '/X', '/Subtype': '/Image', '/Type': '/XObject', '/Width': 890}

Why there is no matrix in the xobject? is it hiding in some other place ,not in XObject?
[jpegcompress.pdf](https://github.com/py-pdf/pypdf/files/12542379/jpegcompress.pdf)

**Comment by pubpub-zz** [comment]
If you use pdfbox in debug (https://github.com/py-pdf/awesome-pdf#file-analysis--security) you can view easily that the image are upside down. Actually the flip in done in pdf using a proper cm matrix. you should see the cm matrix if you use the extract_text() function with a visitor looking at operators and searching for "Do" operators

**Reply by jianfan2012** [reply]
Can you share a sample code for visitor looking for operators?
thanks

**Comment by pubpub-zz** [comment]
As this not an issue I convert it into discussion to ease analysis

**Comment by pubpub-zz** [comment]
@jianfan2012 Can you replace the title with a clearer one

**Reply by jianfan2012** [reply]
@pubpub-zz thank you so much for your quick response ! sure I can replace title , how about "how to extract CM Matrix for images in PDF file " ? 
and Can you show me some sample code of  "you should see the cm matrix if you use the extract_text() function with a visitor looking at operators and searching for "Do" operators"
I can see the cm , and tm by using extract_text(visitor_text=visitor_function)
but I can not tell which cm is for images, is this the right way?
thank you !

**Reply by jianfan2012** [reply]
I think I got it, I use "visitor_operand_before" and "visitor_operand_after" visitor, checking the four arguments "operator, operand-arguments, current transformation matrix and text matrix" , search for "Do" operator, and if "operand-arguments" is  Image XObject name, then it is the image operator

**Reply by pubpub-zz** [reply]
I'm inline with you 👏. Do not hesitate to publish your test results to help others 😋

**Reply by jianfan2012** [reply]
Sample code to get the image current transformation matrix and determine the rotation of images

from pypdf import PdfReader
import PIL
        
page_matrix = {}
def visitor_function(op, args, cm, tm):
       # op : operator
       # args: , operator arguments for image is the XObject name
       # cm: current transformation matrix
       # tm: text matrix

        if op.decode() == 'Do':
             page_matrix[args[0]] = cm

def check_image_rotation(cm):
    """
    :param m: matrix of rotation
    :return: rotated image
    """
    msg = []
    if m[1] * m[2] == 0:  # means rotation by 0, 180, or flippings
        if m[0] * m[3] > 0:  # same sign -> no flippings
            if m[0] < 0:  # so both, a and d are negative!
                msg = [PIL.Image.ROTATE_180]
            else:
                msg = []
        else:  # we have a flip
            if m[0] < 0:  # horizontal flip
                msg = [PIL.Image.FLIP_LEFT_RIGHT]
            else:
                msg = [PIL.Image.FLIP_TOP_BOTTOM]
    elif abs(m[1]) == abs(m[2]):
        if m[1] * m[2] < 0:
            if m[1] > 0:
                msg = [PIL.Image.ROTATE_90]
            else:
                msg = [PIL.Image.ROTATE_270]
        else:
            if m[1] > 0:
                msg = [PIL.Image.FLIP_TOP_BOTTOM, PIL.Image.ROTATE_90]
            else:
                msg = [PIL.Image.ROTATE_90, PIL.Image.FLIP_TOP_BOTTOM]
    else:
        logger.info("unsupported matrix")
    return msg

pdf_file = "jpegcompress.pdf"
doc = PdfReader(pdf_file)
for  page in doc.pages:
      page.extract_text(visitor_operand_before=visitor_function,visitor_operand_after = visitor_function)
for key,cm in page_matrix.items():
     img_rotation = check_image_rotation(cm)
